clang/tools/3c/README.md

```diff
@@ -0,0 +1,65 @@
+
+3. Repeat until you are satisfied.
+
+For a complete, worked example, see [our tutorial](https://github.com/correctcomputation/checkedc-tiny-bignum-c/tree/v0_3cconvert) (FIXME: Update this link once Kyle and Shilpa get things sorted out).
+
+## Basic usage
+
+The task of `3c` is complicated by the fact that the build system for a typical C codebase will call the C compiler once per `.c` file, possibly with different flags for each file (include directories, preprocessor definitions, etc.), and then link all the object files at the end.  A Clang-based whole-program analysis like 3C needs to process all `.c` files at once _with the correct flags for each_.  To achieve this, you get your build system to produce a Clang "compilation database" (a file named `compile_commands.json`) containing the list of `.c` files and the flags for each ([how to do this depends on the build system](../../docs/JSONCompilationDatabase.rst)), and then 3C reads this database.
+
+However, in a simpler setting, you can manually run `3c` on one or more source files, for example:
+
+```
+3c -alltypes -output-postfix=checked foo.c bar.c
+```
+
+This will write the new version of each file `f.c` to `f.checked.c` in the same directory (or `f.h` to `f.checked.h`).  If `f.checked.c` would be identical to `f.c`, it is not created.
+
+As an additional safeguard, `f.checked.c` or `f.checked.h` is not written if it is outside the _base directory_, which defaults to the working directory but can be overridden with the `-base-dir` flag.  This can help ensure that you don't unintentionally modify external libraries, though of course if their header files are not annotated for Checked C, your ability to convert your own program to Checked C may be limited.  (See [below](#annotated-system-headers) about system headers.)  However, there is a bug in the way the base directory check handles `..` path components in `-I` directories and `#include` paths (terrible, we know!), so until we can fix the bug, please avoid using `..` path components (e.g., use `-I` with an absolute path instead).
+
+You can ignore the errors about a compilation database not being found.  You can specify a single set of flags to use for all files by prefixing them with `-extra-arg-before=`, for example:
+
+```
+3c -alltypes -output-postfix=checked -extra-arg-before=-Isome/include/path foo.c bar.c
+```
+
+(If you were using a compilation database, such "extra" flags would be added to any flags in the database.)
+
+The `-alltypes` option causes `3c` to try to infer array types.  We want to make this the default but haven't done so yet because it breaks some things.
```

Comment:
This is a very important, and encouraged option, so I would move it up. Maybe move it above "As an additional safeguard" and then add a `### Other notes` heading above that sentence, since the discussion is a bit more in the weeds.

Reply:
Done.